File: test/google_groups.py
result = None
href = None


def init_results(driver):
    global result
    try:
        result = driver.find_element('partial link text', 'mini-profiler')
    except:
        # Element not found
        pass
    return result


def assert_link_unchanged():
    global result, href
    assert result.get_attribute('href') == href


def assert_no_intermediate_urls(driver, method, target):
    driver.get_urls()
    method()
    urls = []

    def check_urls():
        urls.extend(driver.get_urls())
        return urls
    driver.wait_until(check_urls)
    assert urls[0] == target


def run(driver):
    global href

    # Open Groups post
    driver.navigate('https://groups.google.com/forum/?hl=en' +
                    '#!msg/play-framework/ZfmjuYnZrzg/2hx2zgq_GugJ')
    driver.wait_until(lambda: init_results(driver))

    # Check link URL
    href = result.get_attribute('href')
    assert 'google.com' not in href

    # Right-click the link
    result.context_click()
    assert_link_unchanged()
    result.send_keys(driver.keys.ESCAPE)
    assert_link_unchanged()

    # Click the link
    orig_window = driver.current_chrome_window_handle
    assert_no_intermediate_urls(driver, lambda: result.click(), href)
    driver.close_windows(keep=orig_window)
    assert_link_unchanged()

    # Middle-click search result
    assert_no_intermediate_urls(driver, lambda: result.middle_click(), href)
    driver.close_background_tabs()
    assert_link_unchanged()

    # Click Apps button to bring up dropdown
    driver.find_element('css selector', 'a[title="Google apps"]').click()
    driver.wait_until(lambda: driver.find_element('css selector', 'div[aria-label="Google apps"]').is_displayed())

File: test/google_groups.js
/*
 * This Source Code is subject to the terms of the Mozilla Public License
 * version 2.0 (the "License"). You can obtain a copy of the License at
 * http://mozilla.org/MPL/2.0/.
 */

"use strict";

import * as utils from "../test-utils.js";

const searchResultSelector = "[href*='github.com']";

describe("Google Groups", () =>
{
  let browser;
  let page;
  let href;

  it("should load", async function()
  {
    browser = await utils.launchBrowser();
    page = await browser.newPage();
    await page.goto("https://groups.google.com/forum/?hl=en#!msg/play-framework/ZfmjuYnZrzg/2hx2zgq_GugJ");
    await page.waitForSelector(searchResultSelector);
  });

  it("should not redirect links to Google initially", async function()
  {
    let result = await page.$(searchResultSelector);
    expect(result).to.be.not.null;

    href = await result.getHref();
    expect(href).to.not.include("google.com");
  });


  it("shouldn't change link on hover", async function()
  {
    let result = await page.$(searchResultSelector);

    await result.hover();
    await page.waitFor(200);
    expect(await result.getHref()).to.equal(href);
  });

  it("shouldn't change link on right-click", async function()
  {
    let result = await page.$(searchResultSelector);

    await result.click({button: "right"});
    await page.waitFor(200);
    expect(await result.getHref()).to.be.equal(href);
  });

  it("shouldn't change link on middle-click", async function()
  {
    let result = await page.$(searchResultSelector);

    let target = await utils.newTarget(browser, async function()
    {
      await result.click({button: "middle"});
    });
    expect(target.url()).to.be.equal(href);

    let tab = await target.page();
    await tab.close();

    await page.waitFor(200);
    expect(await result.getHref()).to.be.equal(href);
  });

  it("shouldn't change link on click", async function()
  {
    let result = await page.$(searchResultSelector);

    let target = await utils.newTarget(browser, async function()
    {
      await result.click();
    });
    expect(target.url()).to.be.equal(href);

    let tab = await target.page();
    await tab.close();

    await page.waitFor(200);
    expect(await result.getHref()).to.be.equal(href);
  });

  it("should open apps listing when Google Apps button is clicked", async function()
  {
    let link = await page.$("a[title='Google apps']");
    expect(link).to.be.not.null;

    await link.click();
    await page.waitForSelector("div[aria-label='Google apps']");
  });

  it("should shut down", async function()
  {
    await browser.close();
  });
});
